### ISSUE TITLE
aten::batch_norm schema does not mention mutation of running_mean/running_var

### ISSUE THREAD

**Issue by silvasean** [pytorch/pytorch#73050]
### 🐛 Describe the bug

aten::batch_norm's schema has the following arguments:

```
  arguments:
    arg: {'name': 'input', 'type': 'Tensor', 'pytype': 'Tensor'}
    arg: {'name': 'weight', 'type': 'Tensor?', 'pytype': 'Optional[Tensor]'}
    arg: {'name': 'bias', 'type': 'Tensor?', 'pytype': 'Optional[Tensor]'}
    arg: {'name': 'running_mean', 'type': 'Tensor?', 'pytype': 'Optional[Tensor]'}
    arg: {'name': 'running_var', 'type': 'Tensor?', 'pytype': 'Optional[Tensor]'}
    arg: {'name': 'training', 'type': 'bool', 'pytype': 'bool'}
    arg: {'name': 'momentum', 'type': 'float', 'pytype': 'float'}
    arg: {'name': 'eps', 'type': 'float', 'pytype': 'float'}
    arg: {'name': 'cudnn_enabled', 'type': 'bool', 'pytype': 'bool'}
```

Note that `running_mean` and `running_var` don't have `alias_info`. This makes some sense, since aliases to them are not returned, so there is no need for "alias" info per se, but currently `alias_info` also is the structure that contains the `is_write` flag. This makes it hard for downstream to automatically understand the semantics of the op.

Context: In the [Torch-MLIR](https://github.com/llvm/torch-mlir) project, we [generate](https://github.com/llvm/torch-mlir/blob/ed9bd556b3f59feb0fe1cca9776a3c2eb82fe826/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L1) MLIR operation descriptions (called "ODS") for the operators in the JIT registry. We infer various information and annotate the MLIR operations based on the schema. Two such properties are:
- HasValueSemantics (i.e., is it purely functional). We currently use the absence of alias_info to indicate that, but that seems to be incorrect for aten::batch_norm, which is not purely functional
- ReadOnly (not yet implemented) - a weaker version of HasValueSemantics -- it may introduce aliases, but it is guaranteed not to mutate any operands. Suitable for view-like ops, list getitem, etc.

How can we correctly infer these properties from the registry?

@bdhirsh

### Versions

Currently working off nightly at the time of this bug report.

**Comment by eellison** [comment]
cc @davidberard98 didn't you try to update this and run into some problems ? do you remember ?

**Comment by VitalyFedyunin** [comment]
Marking triaged as answer to the question was provided and we have POC here.

**Comment by davidberard98** [comment]
@silvasean yes, AFAIK, in general the alias annotations are provided by [native_functions.yaml](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml) and other than a few special cases, they should usually be accurate.

Looking at [alias_analysis.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/alias_analysis.cpp#L613) it seems like there are currently special cases only for aten::batch_norm, aten::instance_norm, and aten::wait (which doesn't seem to have an entry in native_functions.yaml anyway)

**Comment by byronyi** [comment]
Should be fixed by https://github.com/pytorch/pytorch/pull/88697.

**Comment by bdhirsh** [comment]
Yep! A bit more context:

(1) This github issue mentions `at::batch_norm` having an incorrect schema. That's technically true, but matters less for backends in practice, because `at::batch_norm` will always decompose into other operators, so backends won't see `at::batch_norm` directly. Today, the common case for other backends is that it decomposes into `at::native_batch_norm`.

(2) `native_batch_norm` also has incorrect schema. That linked PR tries to help by further decomposing `native_batch_norm` into 2 new ops (they're marked as private and hopefully we can give them better names 😄) - one op that takes in running stats, and one that doesn't take in running stats.

The variant that takes in running stats is properly marked with mutable alias annotations its schema. That way, functionalization will see that new op and correctly figure out how to replace it with its functional variant, `_native_batch_norm_legit_functional`. So backends that use functionalization in some way, e.g. through AOT Autograd, will correctly only see functional ops in their graph in a user program involving batch norm.

That PR also includes a decomposition for the new functional op, that backends can opt into.
